src/Elliptic/Systems/Xcts/Equations.cpp

```diff
@@ -0,0 +1,619 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "Elliptic/Systems/Xcts/Equations.hpp"
+
+#include <algorithm>
+#include <cstddef>
+#include <functional>
+#include <optional>
+#include <string>
+
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tags/TempTensor.hpp"
+#include "DataStructures/TempBuffer.hpp"
+#include "DataStructures/Tensor/EagerMath/DotProduct.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "Elliptic/Systems/Xcts/Geometry.hpp"
+#include "PointwiseFunctions/GeneralRelativity/IndexManipulation.hpp"
+#include "Utilities/ConstantExpressions.hpp"
+#include "Utilities/ErrorHandling/Assert.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/Literals.hpp"
+#include "Utilities/MakeWithValue.hpp"
+#include "Utilities/TMPL.hpp"
+
+namespace Xcts {
+namespace detail {
+void fully_contract(const gsl::not_null<Scalar<DataVector>*> result,
+                    const tnsr::II<DataVector, 3>& tensor1,
+                    const tnsr::II<DataVector, 3>& tensor2) noexcept {
+  get(*result) = tensor1.multiplicity(0_st) * tensor1[0] * tensor2[0];
+  for (size_t i = 1; i < tensor1.size(); ++i) {
+    get(*result) += tensor1.multiplicity(i) * tensor1[i] * tensor2[i];
+  }
+}
+void fully_contract(const gsl::not_null<Scalar<DataVector>*> result,
+                    const tnsr::II<DataVector, 3>& tensor1,
+                    const tnsr::II<DataVector, 3>& tensor2,
+                    const tnsr::ii<DataVector, 3>& metric) noexcept {
+  get(*result) = 0.;
+  for (size_t i = 0; i < 3; ++i) {
+    for (size_t j = 0; j < 3; ++j) {
+      for (size_t k = 0; k < 3; ++k) {
+        for (size_t l = 0; l < 3; ++l) {
+          get(*result) += metric.get(i, k) * metric.get(j, l) *
+                          tensor1.get(i, j) * tensor2.get(k, l);
```

Comment:
I actually have some temporary memory to spare at the points where this function is used, so I refactored it to perform 63 multiplications instead of the 243 (63 because the two tensors may not be the same) without allocating more memory.